Implemented Monolog

evrifaessa opened the pull request:
This PR implements [Monolog](https://github.com/Seldaek/monolog/tree/2.x) 2.x. (Will bump to 3.x. once we decide to bump to PHP 8.1 min.).

~~In the future, I'm planning to implement separate channels for different functionality, such as `event`, `http_client`, `cron` and others. Separating the channels would allow us to [log the messages to different files](https://symfony.com/doc/current/logging/channels_handlers.html), resulting in more organized logs.~~ Done.

Monolog can also be adjusted to make use of [different handlers, formatters and processors](https://github.com/Seldaek/monolog/blob/2.x/doc/02-handlers-formatters-processors.md#handlers). It's good.

admdly reviewed (commented):
Generally looks good, but a few errors with the tests need fixing before we can merge.

evrifaessa commented:
An update, I've added channel functionality. The `setChannel` method can easily be chained like this:

```php
$this->di['logger']->setChannel('mail')->debug('foo');
```

This would log the message in the `mail.log` file under the data directory. Any message without a channel will automatically end up in the "application" channel and the `application.log` file.

I've already gone ahead and updated a couple places like the cron jobs and events to get them to use their own channel.

The allowed channels list currently is a static array. I probably will get rid of it in the future and enable any channel to be used.

admdly commented on src/library/Box/Log.php:
```diff
@@ -24,33 +24,32 @@
+    const INFO = 6;  // Informational: informational messages
+    const DEBUG = 7;  // Debug: debug messages
+
+    protected array $_priorities = array(
```
Very minor, but arrays can be declared using the [] syntax since PHP 5.4 which seems to now be preferred. 🙂

admdly commented on src/library/Box/EventManager.php:
```diff
@@ -38,9 +38,7 @@ public function fire($data)
         $subject = $data['subject'] ?? null;
         $params = $data['params'] ?? null;
 
-        if (BB_DEBUG) {
-            $this->di['logger']->debug($event.': '.var_export($params, 1));
-        }
```
Are we always logging debug events now? Or again, has this check been moved? 🙂

evrifaessa replied:
There already is a check here, it's redundant:

https://github.com/FOSSBilling/FOSSBilling/blob/df4dba4c77d917e482fdaaecf40289f96414dd82/src/library/Box/Log.php#L138-L141

admdly replied:
Great - thanks - just what I was looking for! 😄

admdly commented on src/di.php:
```diff
@@ -70,10 +73,6 @@
         }
 
         $log->addWriter($writer2);
-    } else {
-        $logFile = $di['config']['path_logs'];
-        $writer = new Box_LogStream($logFile);
-        $log->addWriter($writer);
```
Do we always now log to disk, regardless of log to DB? Or has this check been moved elsewhere?

evrifaessa replied:
I did that while testing, having them both enabled would be redundant. Reverting that.

admdly reviewed (commented):
Great changes - just a few comments. 👍

BelleNottelling commented on src/library/FOSSBilling/Monolog.php:
```diff
@@ -0,0 +1,109 @@
+        if (isset($map[$priority])) {
+            return $map[$priority];
+        } else {
+            return Logger::DEBUG;
```
Rather than doing an else-if, we can just do this:
`return $map[$priority] ?? Logger::DEBUG;`

BelleNottelling reviewed (commented):
Overall it looks pretty good. Besides what Adam pointed out, I only noticed one thing I'd change
